### ISSUE TITLE
[Feature Request] Access to full controller input

### ISSUE THREAD

**Issue by Mynotaurus** [nesbox/TIC-80#1377]
Just like how key() and keyp() lets us access the full keyboard, could we implement a function that enables us to access every button on a connected gamepad? It could even potentially access analog inputs like joysticks or triggers to add a new aspect of gameplay to games.

**Comment by nesbox** [comment]
Not sure it will be implemented someday, I think you have to use gamepad buttons mapping in the GAMEPAD CONFIG menu to map buttons.

**Comment by RobLoach** [comment]
In addition, what do you mean by "Full Controller"? There are thousands of different controllers out there, all with different button sets, analog inputs, and gyro controls. While I would like to see a Start button added, it can be simulated to the X or Y buttons, or combinations.

**Comment by Mynotaurus** [comment]
Good point. I was thinking there would probably be difficulties in terms of different controllers having different inputs, but I was unsure of how TIC-80 parses controller input currently.

**Comment by joshgoebel** [comment]
Well the problem also is that you only have so much "space" for the controller data at the "hardware level"...  See the RAM map:

https://github.com/nesbox/TIC-80/wiki/RAM

Currently each controller has just 8 bits (which are all already used)... so that means ALL you can track (without making a newer "incompatible" version of the "hardware") is 8 buttons... so that also would have to be solved to allow more controller input.

**Comment by Mynotaurus** [comment]
Ahh, I didn't realise the keyboard was being read from RAM, I can see why this would be so difficult now.

**Comment by joshgoebel** [comment]
@nesbox Shouldn't we perhaps rope off some RAM as at least listed as "reserved" for future use, or is the "hardware" truly mostly done evolving... or if you added features later you'd just carve away from "free" space, making games that used that space incompatible?  Right now as I count we have like 3 bytes of reserved RAM (not counting the bits here and there in other data fields).

**Comment by nesbox** [comment]
We use RAM to talk with "hardware" only and people don't usually use it to store data because they can create as many variables as they want in fantasy consoles.
I don't think so we should rope off this "free" space.
Thanks.

**Comment by joshgoebel** [comment]
> I don't think so we should rope off this "free" space.

Well it's obviously your call. :-)

> We use RAM to talk with "hardware" only and people don't usually use it to store data because they can create as many variables as they want in fantasy consoles.

_But I'd disagree with this entirely in my personal experience._ 

I think the "hardware" RAM is a hugely fun limitation to have at ones disposal - much like all the other limitations.  Why does switching banks matter at all - why is it so limited?  Why can't someone just page all 8 banks into "variables" when the cartridge boots?  They could - **but that's kinda lame.**  It's way more fun to write a renderer that renders from 1.5 banks by loading an additional portion of sprites into high RAM (the 12kb that's free) and then either have a custom render function or swap that sprite data in and out of SPRITE ram. (perhaps for parallax scrolling, layers, etc)

I almost did exactly this for a game I was working on for a recent jam just to get access to an extra 128 sprites without "cheating". Just paging tons of banks into the Lua VM would be cheating, but using the extra RAM the "hardware" provides is just thinking creatively and I imagine sort of like all the crazy trick people came up with to make early hardware do things that people didn't think it could do. :)
